File: Toal/scales.py
# This C99 application takes a single command line argument which is the name of
# a note and prints, to standard output, the major (Ionian) and minor (Aeolian)
# scales starting at that note.  The program is very weak; note names are
# constrained to be in {A, A#, B, C, C#, D, D#, E, F, F#, G, G#}. You should
# totally enhance this to use flats properly, and even double sharps and double
# flats.

import sys

KEY_NAMES = ['A', 'A#', 'B', 'C', 'C#', 'D', 'D#', 'E', 'F', 'F#', 'G', 'G#']
NUMBER_OF_NOTES = len(KEY_NAMES)

MAJOR_STEPS = [2,2,1,2,2,2]
MINOR_STEPS = [2,1,2,2,1,2]
NUMBER_OF_STEPS = len(MAJOR_STEPS)

def print_scale(note_index, type, steps):
    print(f'{KEY_NAMES[note_index]:3}{type}: {KEY_NAMES[note_index]:3}', end='')
    offset = 0
    for step in steps:
        offset += step
        print(f'{KEY_NAMES[(note_index + offset) % NUMBER_OF_NOTES]:3}', end='')
    print()

if __name__ == '__main__':
    if len(sys.argv) != 2:
        print("This program requires exactly one command line argument\n")
        sys.exit(1)

    try:
        key_index = KEY_NAMES.index(sys.argv[1].upper())
    except ValueError:
        print(f'No such key: {sys.argv[1]}')
        sys.exit(2)

    print_scale(key_index, "major", MAJOR_STEPS)
    print_scale(key_index, "minor", MINOR_STEPS)

File: Toal/scales.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <math.h>

char *KEY_NAMES[] = {"A", "A#", "B", "C", "C#", "D", "D#", "E", "F", "F#", "G", "G#"};
int NUMBER_OF_NOTES = 12;
int MAJOR_STEPS[] = {2,2,1,2,2,2};
int MINOR_STEPS[] = {2,1,2,2,1,2};

int indexOfNote(char* note) {
  int noteNum;
  for(int i = 0; i < NUMBER_OF_NOTES; i++) {
    if (strcmp(KEY_NAMES[i], note) == 0) {
      noteNum = i;
    }
  }
  return noteNum;
}

void print_scale(int note_index) {
  printf("%s major: %s ", KEY_NAMES[note_index], KEY_NAMES[note_index]);
  int offsetMajor = 0;
  for (int stepIndexMajor = 0; stepIndexMajor < 6; stepIndexMajor++) {
    offsetMajor += MAJOR_STEPS[stepIndexMajor];
    printf("%s ", KEY_NAMES[(note_index + offsetMajor) % NUMBER_OF_NOTES]);
  }
  printf("\n%s major: %s ", KEY_NAMES[note_index], KEY_NAMES[note_index]);
  int offsetMinor = 0;
  for (int stepIndexMinor = 0; stepIndexMinor < 6; stepIndexMinor++) {
    offsetMinor += MINOR_STEPS[stepIndexMinor];
    printf("%s ", KEY_NAMES[(note_index + offsetMinor) % NUMBER_OF_NOTES]);
  }
}

int main(int argc, char *argv[]) {
  print_scale(indexOfNote(argv[1]));
  return 0;
}
